Reconfigure task rate limits

king-alexander opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
This pull request turns the `summary_by_domain` task rate limit into a configuration option. 
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

- Resolves #32 

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

I tested these changes by configuring the admiral `cert-workers` to run with 2 different rate limits in place. The requests made to the [Certificate Transparency Search API](https://sslmate.com/ct_search_api/) were expected to succeed in one case and fail in the other. 

By default, the API permits a maximum of 10 requests per hour for full-domain certificate queries. In order to meet this limit, I had to account for the way in which Celery applies rate limits for tasks. Rate limits are applied on a _per-worker_ basis, so with 5 `cert-worker` replicas, I adjusted the rate limit to 2 requests per hour. This produced a global rate limit of 10 requests per hour. From there, I verified that requests completed successfully.

For the other case, I changed the rate limit to 1 request per second. As expected, this failed. The rate limit triggered HTTP 429 errors for generating too many requests.

See the screenshots below for the result of each case respectively.


## 📷 Screenshots ##
<img width="658" alt="Screen Shot 2023-01-31 at 1 44 12 PM" src="https://user-images.githubusercontent.com/55767786/215866047-74918c70-1efa-425e-974b-f0fecfc44916.png">
<img width="657" alt="Screen Shot 2023-01-31 at 1 50 24 PM" src="https://user-images.githubusercontent.com/55767786/215867350-e4029575-58ee-437a-a68b-523aae5600e6.png">

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
<!--
- [ ] *All* future TODOs are captured in issues, which are referenced
      in code comments.
-->
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
<!--
- [ ] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
-->
<!--
- [ ] Tests have been added and/or modified to cover the changes in this PR.
- [ ] -->
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been approved. -->

- [x] Revert dependencies to default branches.
- [x] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Add a tag or create a release.

king-alexander committed:
Remove summary_by_domain default rate limit

king-alexander committed:
Alphabetize celery-defaults section

king-alexander commented:
I would like to get more experienced eyes on these changes. Is there a better way to do this?

dav3r commented on secrets/admiral.yml:
```diff
@@ -37,9 +37,38 @@ dev-mode:
+  autodiscover_tasks:
+    - admiral.certs
+
+cert-worker-large:
   celery:
     <<: *celery-defaults
+    # 1000 full-domain queries per hour
+    task_annotations: {'tasks.summary_by_domain': '1000/h'}
```
Instead of making new `cert-worker-small`/`cert-worker-medium`/`cert-worker-large` sections, why not just make the rate limit itself the configurable item?

```suggestion
task_rate_limit: &task_rate_limit
  # Modify the rate below as needed.  "10/h" means limit to 10 requests per hour.
  summary_by_domain: 10/h

cert-worker:
  celery:
    <<: *celery-defaults
    task_annotations:
      tasks: *task_rate_limit
```

This allows you to remove the comments above that are specific to the various sslmate tiers and keep everything more generalized to the number of requests per hour.

king-alexander replied:
Thank you Dave! I incorporated your suggestion in [d59c340](https://github.com/cisagov/admiral/pull/37/commits/d59c340d0fac65a6c2bd6b0a114d7c3b2eb2e5c3). I had to make a few minor changes to put the configuration in the format Celery expects.

dav3r replied:
Looks great! 👍

dav3r reviewed (changes requested):
Please take a look at my comment and let me know if you have any questions.

king-alexander committed:
Make task rate limit configurable

With this change, we can limit the number of requests performed within a
given time frame. We can specify seconds, minutes, or hours over which
tasks will be evenly distributed.

dav3r commented:
Don't forget to update the testing section of this PR after you've tested the changes in https://github.com/cisagov/admiral/pull/37/commits/d59c340d0fac65a6c2bd6b0a114d7c3b2eb2e5c3.

dav3r commented:
@king-alexander Did you confirm that varying the rate limit worked as expected?  i.e. did you test `1/h` vs. `10/h` and verify that Celery enforced the correct rate?

dav3r reviewed (commented):
@king-alexander Did you confirm that varying the rate limit worked as expected? i.e. did you test `1/h` vs. `10/h` and verify that Celery enforced the correct rate?

king-alexander commented:
> @king-alexander Did you confirm that varying the rate limit worked as expected? i.e. did you test `1/h` vs. `10/h` and verify that Celery enforced the correct rate?

I can confirm. As expected, when I changed the rate limit to `1/s` I received multiple HTTP 429 Errors. I forgot, however, that the rate limit is applied _per worker_, which means to produce a global rate limit we must multiply the rate limit put in place by the number of `cert-workers` we provision. I made [02cf0ce](https://github.com/cisagov/admiral/pull/37/commits/02cf0ce988ec0a9a34057a6acf3549185d6aceb7) to address this.

king-alexander committed:
Establish new default rate limit

This commit establishes a new default rate limit for the
summary_by_domain task. Since the rate limit is applied per worker, I
reduced the limit to 2 tasks per hour and the number of cert-worker
replicas to 5 (for easy math). This way, we can match the rate limit for
the Small Tier of the Certificate Transparency Search API.

dav3r reviewed (approved):
Looks great @king-alexander. 🚀   

One last request- can you please update the "Testing" section of this PR to document a bit more detail about what and how you tested?

dav3r commented on secrets/admiral.yml:
```diff
@@ -37,9 +37,17 @@ dev-mode:
     - admiral.tester
 
+task_rate_limit: &task_rate_limit
+  # Modify the rate below as needed. Limits are applied per worker, so
+  # "2/h" means limit one worker to 2 tasks per hour. To get the global
+  # rate limit, you must multiply by the number of workers provisioned.
+  rate_limit: "2/h"
```
I just thought of one other thing that I'd like to see here.  Can this comment please include a link to the Celery documentation where the rate limit syntax is defined, in case someone wants to use minutes or seconds, rather than hours?  I think [this is the right link](https://docs.celeryq.dev/en/2.4-archived/reference/celery.app.task.html#celery.app.task.BaseTask.rate_limit), but please confirm.

king-alexander replied:
See [5875531](https://github.com/cisagov/admiral/pull/37/commits/58755315a87eea9c50ed11c21532e397bea0f762). That is the right link, but I included a reference from the Celery User Guide instead because it has all the pertinent information in one place. It provides syntax examples and notes that rate limits are applied on a per-worker basis.

dav3r replied:
Good call- the link you used is definitely better.  Thanks!

jsf9k reviewed (approved):
Approved, once @dav3r's last request is addressed.

king-alexander committed:
Include rate limit syntax reference

dav3r commented:
@king-alexander I see you updated the "Testing" section with more detail, but it's not clear to me how the first and third tests are different.  They both appear to be 10 requests per hour.

king-alexander commented:
> @king-alexander I see you updated the "Testing" section with more detail, but it's not clear to me how the first and third tests are different. They both appear to be 10 requests per hour.

Thanks for pointing this out! I'm updating the section to clarify.

I described my testing in terms of how it unfolded rather than the significant results. The first run wasn't important to testing because I had the rate limit wrong. That was 10 requests per hour for each worker, not 10 requests per hour total like it should've been. Anyway, I hope `Testing` is easier to reason about now.

dav3r commented:
> > @king-alexander I see you updated the "Testing" section with more detail, but it's not clear to me how the first and third tests are different. They both appear to be 10 requests per hour.
> 
> Thanks for pointing this out! I'm updating the section to clarify.
> 
> I described my testing in terms of how it unfolded rather than the significant results. The first run wasn't important to testing because I had the rate limit wrong. That was 10 requests per hour for each worker, not 10 requests per hour total like it should've been. Anyway, I hope `Testing` is easier to reason about now.

Thanks for updating your testing narrative- it's much clearer now. 👍

king-alexander committed:
Bump version from 1.2.0 to 1.3.0
